viscy/representation/disentanglement_metrics.py

```diff
@@ -0,0 +1,374 @@
+import logging
+from typing import Dict, Optional, Tuple
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from scipy import stats
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.linear_model import LogisticRegression
+from sklearn.metrics import accuracy_score
+
+_logger = logging.getLogger(__name__)
+
+
+class DisentanglementMetrics:
+    """
+    Disentanglement metrics for VAE evaluation on microscopy data.
+
+    Implements MIG, SAP, DCI, and Beta-VAE score metrics for evaluating
+    how well the VAE learns disentangled representations.
+    """
+
+    def __init__(self, device: str = "cuda"):
+        self.device = device
+
+    def compute_all_metrics(
+        self,
+        vae_model: nn.Module,
+        dataloader: torch.utils.data.DataLoader,
+        max_samples: int = 1000,
+        n_factors: Optional[int] = None,
+    ) -> Dict[str, float]:
+        """
+        Compute all disentanglement metrics.
+
+        Args:
+            vae_model: Trained VAE model
+            dataloader: DataLoader with labeled data
+            max_samples: Maximum number of samples to use
+            n_factors: Number of known generative factors (if available)
+
+        Returns:
+            Dictionary of metric scores
+        """
+        latents, factors = self._extract_latents_and_factors(
+            vae_model, dataloader, max_samples
+        )
+
+        metrics = {}
+
+        # MIG Score
+        metrics["MIG"] = self.compute_mig(latents, factors)
+
+        # SAP Score
+        metrics["SAP"] = self.compute_sap(latents, factors)
+
+        # DCI Scores
+        dci_scores = self.compute_dci(latents, factors)
+        metrics.update(dci_scores)
+
+        # Beta-VAE Score (unsupervised)
+        metrics["Beta_VAE_Score"] = self.compute_beta_vae_score(
+            vae_model, dataloader, max_samples
+        )
+
+        return metrics
+
+    def _extract_latents_and_factors(
+        self,
+        vae_model: nn.Module,
+        dataloader: torch.utils.data.DataLoader,
+        max_samples: int,
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Extract latent representations and generative factors.
+
+        For microscopy data, we'll extract simple visual factors like:
+        - Cell size (approximated from pixel intensity)
+        - Cell count (approximated from connected components)
+        - Brightness (mean intensity)
+        - Contrast (std of intensity)
+        """
+        vae_model.eval()
+        latents = []
+        factors = []
+
+        samples_collected = 0
+
+        with torch.no_grad():
+            for batch in dataloader:
```

Comment:
Is this running the same validation epoch twice?

Reply:
removed this because I can't test it now. If we need this I'll add it later on in a separate PR